### DISCUSSION TITLE
Long note custom visual length adjustment for mania (percy/cut long notes)

### DISCUSSION THREAD

**Discussion by Endroma1** [ppy/osu#20576] [Ideas]
As percy/cut long note skins with custom long note ends becomes more and more common in osu mania, it would be nice to support this in lazer. With the upcoming long note calculation rework it will become more justifiable to implement this directly in to skinning/layout editor as long notes will be less powerful.

Implementing this would increase the willingness to switch from stable as switching back to non-percy is quite disencouraging. Especially for newer mania players that simply find top player skins with percy LNs.  I would personally never do the change as long as lazer don't support it or if stable continues the support. There is little to no reason to change from stable when you would lose performance doing so.

It is worth mentioning that percy/cut long notes are still somewhat controversial but i haven't managed to find a lot of discussion around it.

**Comment by smoogipoo** [comment]
I have no idea what this is. Can you provide examples, skins, etc?

**Reply by peppy** [reply]
It's this thing:

https://github.com/ppy/osu/issues/20295
https://github.com/ppy/osu/issues/12057

It's on my list of things to investigate and attempt to understand, but I completely agree that documentation surrounding these skins and what they do / why they have 40,000px images is beyond my comprehension as of yet.

I get the feeling it's obfuscated intentionally as it seems to give some gameplay advantage, but we should understand it and either officially support it in a better way (with backwards compatibility support) or drop support with good rationale.

More example skins and videos would be appreciated regardless (there's one linked in the above issues though).

**Reply by Walavouchey** [reply]
a percy LN is something like this, a hold note with its tail either faded out or cut short. these apparently can make reading easier

![image](https://user-images.githubusercontent.com/36758269/193984144-e4a82d9f-5e91-4636-9d0f-37c17a0f4f63.png)

[skin](https://mega.nz/#!dOZxlQgA!9hU7wLeu_UMKgQXUF9wMXXTaJmznYPk25lGta-7061U)
[different skin](https://www.mediafire.com/file/2y5jhpqwipdh5ry/R_Skin_3.0_monkaLN2_%2528Jakads_%252B_Greatvictory_%252B_LunarCoffee_edit%2529.osk/file)

these particular ones at least don't work in lazer

**Reply by peppy** [reply]
And you know where it ends due to the 1px thick line?

![Safari 2022-10-05 at 05 10 48](https://user-images.githubusercontent.com/191335/193985930-4788a0fc-4f08-4a3b-9403-3fce59160f46.png)

 How do you know when to let go?

**Reply by Walavouchey** [reply]
you can tell by length of the fade and the rhythm. a bit like hidden in the osu mode

if the hold notes have predictable lengths, then you can essentially treat the map as only having regular notes because of the reduced visual clutter. dense maps with a lot of short hold notes sprinkled in become easier to read this way if the release timings are all consistent anyway.

**Reply by Endroma1** [reply]
For most people (such as me) the long notes are too long and i always used to hit too late. Percy lns fixed this for me. Some might like shorter cut and some are used to a longer cut. This is only a visual thing and the discussion is generally about if changing things visually really is abuse/cheating. 

I think it is mostly the method that i used to achieve this that make it feel like cheating. If it were to be implemented "officially" it may feel different and it wouldn't be an gameplay advantage as it would be more easily available. 

It isn't like the taiko-mania situation where the skin changed the gamemode completely. I would still say this is in the realm of skinning elements to make gameplay look cleaner. It might be a little step further than hiding 300s (300gs in mania) by replacing the hit300 element with an empty sprite but it is still the same concept.

**Reply by ikzune** [reply]
This type of skinning typically allows people to play long note dan courses that are 1-2 levels above their standard so it should be a mod allowing us recognize when it was used or not and so people can prevent it's use in certain scenarios such as tournaments that don't want this.

It would be best if the mod had 2 variables:

1. Cut position - how many ms before the end of a hold note does it cut off
2. Fade position - how many ms before the end of a hold note does it start to fade

**Comment by Doksha** [comment]
This seems to be more of a bug with lazer just not properly mirroring stable skins for mania in this specific instance. I don't get why there would be any controversy, some standard players play with dimmed slider bodies for decluttering, why shouldn't mania be able to do the same?
Its more preference than an actual advantage.

**Reply by ghost** [reply]
it looks like the issue is more of technical limitations and less of playability concerns

**Reply by ger0** [reply]
This isn't like dimming the long note body though. You can dim the LN body and it will reduce clutter as well, the problem with percy LN is that it changes the length of a hold object, basically it can reduce 1/8, 1/4 or sometimes even longer beat longnotes into a single short note. This esentially changes the way patterns in the notechart are to be read and played by the player into a different, simpler ones. It can completely remove the original patterns the mapper intended to map (reducing reverse longnote walls to a shortnote pattern with only a few longnotes visible - instead of reading the gaps between the objects and pulling your fingers upwards when the right time comes you simply press it like a normal note and release way earlier, ignoring the supposed length of an object). In comparison to osu!standard this would be similar to making a skin which would reduce the slider complexity so that the sliderbody only goes through the ticks ignoring the original curves mapped by a mapper.

While I understand that the devs do not care about changing the meta in how the patterns are being played, from a mapper perspective it simply hurts to see your original creation being butchered and simplified by a visual change that is in reality just almost the same as if you modified an original notechart.

**Reply by Endroma1** [reply]
If you press certain long notes as shortnotes, wouldn't it make sense for them to look similar? It felt a little bit weird when i started playing mania that certain long notes look so long when they act as short notes. 

Percy lns definetively breaks loose ln wall patterns but not the ones that are dense enough. Although, this comes with the question if mappers should map with percy lns in mind or not. I don't know if this will break the current version of the pp calculations for long notes but it is certainly a point of concern. It would also break certain intended ln patterns on maps not mapped with percy in mind and would almost certainly force ln mappers to use percy lns. 

I think the comparison to standard slider isn't completely fair as percy lns doesn't change any properties of the long notes, only the visual length. It would be more similar if the sliders in standard became visually shorter or faded so dense sliderspam maps become more readable.

**Comment by peppy** [comment]
Proposing a compromise: we allow a setting (and automatically apply when a skin has a 40,000px texture that would otherwise not display) where the hold note body can fade towards the end point, but not hard-cut-off at a ratio of its length. Would this satisfy players to some extent from both parties?

I think at very least this would appease the perspective of a mapper being annoyed at users playing their beatmap in a visually unintended way.

**Reply by Endroma1** [reply]
The problem with not implementing percy lns is that a lot of top scores are set with these lns already. I have seen some people that play with percy lns only when playing ranked maps. This would completly break the leaderboards in term of comparisons between newer and older scores. I think ikzunes solution above is much better as it will separate plays when percy lns are not desired (like in certain dan courses). This would also separate more impressive non-percy plays from not-so impressive percy plays. Not implementing it properly will make it much harder to convince the playerbase that play with percy lns to switch to lazer. 

Custom sliderends is its own thing, as faded slider ends is just a variation of skinning the long-note end. I don't use faded but rather pointy ends as i like it better. Faded, pointy and diamond ends are all popular and implementing faded would only satisfy a part of the playerbase. Making a custom long-note-end skinning element would be more optimal.

**Reply by peppy** [reply]
The fade would be different from what can be done at a skin level (similar to the hack being discussed in this thread), so please understand that.

As for leaderboard concerns: we've made compatibility changes in the past and are not against doing so in the present or future with the idea that it "breaks" them. This is a dynamic ecosystem and for users that care they could check the date at which the score was set and gather what they want from that.

**Reply by caffeine-fighter** [reply]
Don't agree with this, you're not considering the majority of players who use circles, diamonds, or other sliderend images. It also discourages beginners to practice ln accuracy. Encouraging custom skin would be a better solution. Mappers shouldn't force players to play as they intended.

**Reply by peppy** [reply]
In case there is any confusion, this would be my proposal (with tail notes on the left, without on the right):

<img width="553" alt="2022-10-21 08 55 24@2x" src="https://user-images.githubusercontent.com/191335/197080743-c87742e3-63a4-4cd6-9d5a-4027cff9506c.png">

Note that:

- Gradient is applied over the full length (hold nodes will still be >0% opacity until the end, but drop off).
- Hold note overall visual structure does not changes as a result, but noise is reduced somewhat.
- This cannot be done currently at a skin level because it requires knowledge of the hold note duration.

**Reply by Endroma1** [reply]
The left proposal seems good and is quite similar to the gradiant lns that already exist in stable. Wouldn't really call them cut/percy lns but it does help with release timings. 

The right proposal looks straight up like cut/percy lns. The sliderend (the blue part) is removed and it looks physically shorter. 

The bigger problem with cut lns is that it gives such a huge advantage so players can get more than a 5-10% acc improvement on a map. Gaining 10% more pp by just changing a skin does have a huge impact on leaderboards. Nothing in other gamemodes have been able to do such a big change except for taikomania (and maybe touchscreen). It is often called cheating because it gives such a big advantage although, it isn't an unfair one as everyone has access to it. The big problem is if it deviates from normal gameplay. It does change the gameplay as some lns become shortnotes but it doesn't really change the core gameplay of mania. Although, you can say it changes the core gameplay of vsrgs as it will be different from how other vsrgs are played.

If percy lns is implemented, players will adapt to changes and leaderboards will change with it. But scores set without percy lns will be worth less although they are more difficult. It is questionable to say if this is genuine to vsrgs.

If percy lns will not be implemented in lazer but stay in stable it will certanly discourage the players that are used to percy lns to change over. By not fixing percy lns ealy, the younger playerbase that has gotten used to them will be especially discouraged. 

If percy lns will not be implemented in lazer and patched in stable, it will force the whole playerbase to use normal lns. It will certainly discourage some players and make some people angry but it will silence the problem of cut/percy lns completly. 

I still think ikzunes mod proposal is the better option, but if it is not possible, i can't really recommend implementing percy/cut lns.

**Reply by peppy** [reply]
The only difference between the two (left and right) is whether there is a tail note or not. I believe this is not going to be something we should stop (or can stop). Some users will want tail note sprites, others won't.

Anyways, I'm interested in more opinions on the matter because so far this is the best path forward I can imagine.

**Reply by ikzune** [reply]
Current discussions on discord are showing that in extreme cases the difference between a skin that fades and one that doesn't is a B to S rank but in general there is a inherent accuracy and score advantage to using this type of skin. This is because of two reasons, 1. Players naturally release hold notes late and in the right example you provided, even though the top is still visible but faded, it allows players to time their releases better 2. by not paying attention to the extreme fades of each note, players have more time to react to the following notes, inverse typically gives you double the time to react taking advantage of a fade skin.

The prevailing proposals are a harsh modifier maybe 0.75-0.85 to discourage the use of fades or a light modifier maybe 0.85-0.95x pp and/or score pp multiplier to encourage the use of fade as a tool when it is advantageous to do so. This would need to be decided by the pp committee as they would have a better idea on correct multiplier for balancing purposes. but it wouldn't be implementable until lazer is required I think due to stable allowing for this to be skinned.

Regardless of how fade is implemented players who solely play without these skin elements don't want to have to compare to players who are taking advantage of them.

**Reply by peppy** [reply]
You are replying to my proposal with zero mention of the proposal so I'll hide your reply for now. Please post elsewhere (ie. at a top level) to keep things organised better, thanks 🙏 

---

A brief reply to your post though:

PP changes are definitely out of question, unless it is moved to a mod. But that's a whole new discussion. This one is to bring back some level of compatibility via a compromise or otherwise.

I hope more people will arrive in this thread which understand the history of the issue, how skinning works on osu!, and can take an objective approach to it.

**Comment by adaff2** [comment]
`Honest review from <100 ranked player:

1. What does a percy skin do ?
_Percy skins are used to visually shorten the length of long notes. Read above, it is already explained._

2. Does percy skin make reading easier ?
_It depends on the map and overall on the difficulty of the map. For low level maps that require high accuracy on long notes (i.e. 1% of the maps and virtually 0% of the total pp mass), percy will be a disadvantage as it will make it harder to get a good accuracy, particularly making it difficult to judge where exactly the long note really ends.
For high level maps with high long note density, percy has a **significant** impact on the way one reads the notes, making it way easier to both READ and HIT the long notes. It works on hard maps because players tend to hold the long notes for too long by default when the density is high, with percy counterbalancing this tendency and thus making it so that the player hits perfectly or close to it. This leads to score improvements that can go up to 6% accuracy wise (for example a 87% score with a regular skin, and 93% with a percy skin) in extreme cases, and perhaps an improvement of 1%-1.5% on average_ 

3. Is it cheating ?
_It depends on what you consider to be "cheating". If "cheating" is the access to "an advantage over other players" (which is apparently what people had in mind in the Twitter poll), then it is not cheating as everyone has access to it. Well then taiko-mania isn't cheating either, so that's not a valid point.
I don't like either how the question is phrased; what I can say though is that unlike other skin features (note width, note height, column start, hitposition, colours, ...) percy skin gives a significant advantage and drastically changes the gameplay **for everyone**, whereas other skin features are just preferences that might help **some** (depending on the feature) players to feel more comfortable and **slightly** improve their reading. But not everyone wants to use percy skins, for it is first not very aesthetically enjoyable neither fulfilling dignity wise, so you have to understand some players are not willing to use such skins. 
Long notes are a skillset like any other. It would be silly to have a skin making say chordstreams (somehow) way easier to both read and hit, and if it did exist people would be outraged and call it cheating. Percy skins overall are unfair for the (very large) long note community that enjoys playing original LNs and that has been playing and practicing for years before seeing their skillset being singlehandedly destroyed by a type of skin. Hence players (including me) getting discouraged about the state of the game and the ranked maps, players moving back to o2jam, etc._

I have to note that over debating with various players on this subject, the rank (or overall skill level) of the player had somewhat of a weight in their opinion, and if I had to make a remark, it would be that low rank players tend to underestimate the impact of the percy skin on both in game scores and the game overall. Certainly because percy skins have less of an impact on low level maps, and also surely because the small positive impact it has on their scores gives low level players the impression of a fast progression in the game, impression they're understandably not ready to give away.

I understand long notes are difficult and may be discouraging for players as a skillset to learn. But as percy skins slowly made their apparition, I can affirm that players that had been practicing this skillset for years and that had come close to mastering, really took it as an insult to their skillset they had been hardworking on.

4. About potentially implementing
The only reason to implement it in Lazer would be to add it as a mod (similarly to the mirror mod or the hidden mod) and have its own score/pp multiplier. It is absolutely clear that percy skins make the game easier (otherwise there wouldn't be such hype around the skin and no one would ask for it to be implemented in the first place), and mods that make the game easier, like the "easy" mod have a <1 multiplier. I can lead my own research on how much on average percy skins improve players' scores depending on map difficulty, therefore naturally giving a fair value for what the multiplier should be.
The objective here would be that any player should reach more or less the same score whether they're playing with the percy mod (with a <1 score multiplier) or with a normal skin, leaving everyone free to choose for each map, if they think they'd obtain a slightly better score with or without percy mod.

Please consider these as well:
[Percy vs non percy](https://www.youtube.com/watch?v=LES_3QJU3D4)
[Jakads (rank 1 player) reaction to percy](https://youtu.be/HWjPBev10Tg?t=5285) (there's a timestamp)


TL;DR : Percy skins do have a significant impact on gameplay making the game overall way easier, and if it were to be implemented it would need a <1 score multiplier.
As for the 4k/7k difference, I personally think that 4k LN isn't as much of a skillset as 7k LN is so my whole comment really holds only for 7k.

**Comment by andrewletz** [comment]
Please do not discontinue support for 40,000px sprites in lazer. They are not used exclusively for percy/LN cut skins, but also for implementing a gradient on the LN hold body that changes color depending on the length of the long note. This is very useful visual information to have and is only possible through supporting super tall sprites.

[Here is a quick video I made showcasing my skin which uses long note gradients.](https://www.youtube.com/watch?v=nWCtI3yUTR0) 
(From the longest duration long note to shortest it goes, Red -> Purple -> Blue -> Gray -> Cyan tip on the release)
[Here is an imgur link to the hold body itself so you can see how the gradient looks in its full glory.](https://i.imgur.com/24f8Uxj.png)

It is not quite 40,000px tall but 38,600px (an arbitrary number I chose). I know that this is a very niche thing and not many people care but  I would be devastated if gradient LNs weren't supported in lazer. I think this has potential to catch on or even be used in a default skin as it is very helpful for telling which notes you need to hold during LN heavy patterns, especially in inverse maps, but in comparison to percy this doesn't fundamentally alter the appearance and reading of LN patterns.

**Reply by peppy** [reply]
You can do this using a much shorter texture. So please do. 2048px should be a safe value.

**Reply by andrewletz** [reply]
Sadly when the texture is too short it begins repeating throughout the long note, I actually had to scale it up from a 20,000px height in my original version because I was already facing that issue in several long note maps. I can't really think of a workaround that would make gradient hold bodies possible with a 2048px tall texture. For example if it were stretched to avoid the repeating issue it would make the texture inconsistent, like the cyan hold end at the very top of the texture would be stretched, and the length of each colored section would be inconsistent.

**Reply by Walavouchey** [reply]
Since the texture isn't stretched (apparently you can choose whether to stretch or repeat in the skin.ini), maybe lazer could support a numbered hold body sprite, e.g. `mania-hold-#` (perhaps with some other naming scheme if you want to go as far as to support animating it), to essentially split the otherwise tall texture. Alternatively, very tall sprites could be split internally.

This would work around GPU texture size limits while also providing the potential to reduce memory usage when very long hold notes aren't present.

**Reply by andrewletz** [reply]
Splitting the texture internally is a great idea! Thank you for the input.

**Reply by peppy** [reply]
This is not an option. If we are actually supporting stupid tall textures, then it is automatic and implemented for legacy skin support only. It would not be added as some new method/filename/texture/whatever.

**Reply by andrewletz** [reply]
Well, if there is no easy workaround, I would greatly appreciate an implementation to support stupid tall textures. While it isn't common I'm definitely not the only person to use a extremely tall hold body texture as it is the only way to reasonably skin a long note to change throughout the duration of the hold. It would definitely be a sad moment if Lazer couldn't support skinning mania like this. Like I said in my original comment as well I think this has potential to be very well received if implemented in the default mania skin, more visual information when playing is always a good thing in my opinion.

**Comment by longnguyen2004** [comment]
My 2 cents on the issue:

I think we can all agree that 40k textures are a hack, and practically no GPU supports it natively.

Direct3D 11 limits: `D3D11_REQ_TEXTURE2D_U_OR_V_DIMENSION = 16384`
(Source: https://learn.microsoft.com/en-us/windows/win32/direct3d11/overviews-direct3d-11-resources-limits)

OpenGL limits:
| Ryzen 5 5600H Integrated Graphics | NVIDIA GeForce GTX 1650 |
|-|-|
| ![image](https://user-images.githubusercontent.com/15729831/215746652-ebc781cd-ad1c-4ed2-8a28-64187ab8e5df.png) | ![image](https://user-images.githubusercontent.com/15729831/215746850-b401ffca-b063-4d45-9355-9ac09c473879.png) |

(Source: https://www.geeks3d.com/glz/)

So, instead of trying to make huge textures work (which is completely not how GPUs work, and would increase code complexity), I suggest we allow `mania-note*T` to have the top as its origin (controllable through `skin.ini`), which would allow the tail to overlap the body. Since most usage of huge textures (including Percy LNs) I've seen fall under that category, this change would satisfy the majority of the player base. Of course, this would break forward compatibility between stable and lazer, but I think it is a reasonable solution to implement.

**Reply by longnguyen2004** [reply]
https://github.com/ppy/osu/blob/7d1b5751fdb4428da009e6cb9ffa2ef143723a6e/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNoteTail.cs#L34-L39

If I'm reading this right, origin of the tail is already at TopCentre, so it should already be good to go, or maybe I'm missing something...

**Reply by longnguyen2004** [reply]
https://github.com/ppy/osu/blob/9dadd39a2ceb7df91ba68d33c0b4816784a9c7c1/osu.Game.Rulesets.Mania/Skinning/Legacy/LegacyNotePiece.cs#L69-L81
Missed this one here, and also a few other processing logic around that..

**Comment by staticssleever668** [comment]
With new lazer releases featuring score submission, and stable scores now being be capped lower, this is the perfect moment to end the percy/cut/cheat-ln meta.
See https://github.com/ppy/osu/discussions/20576#discussioncomment-3938241 and consider reverting the support for absurdly long textures.
If a score was set with such skin, it should at least be indicated as such, as it's completely unfair in comparison to normal plays. I don't understand how this can be seen as anything but cheating, as this is far beyond simple abusing. Everybody would say so, wasn't it normalized as the osu! meta now.

**Reply by longnguyen2004** [reply]
> stable scores now being be capped lower

This is not the final value, and will most likely be set back if weirdness happens

> consider reverting the support for absurdly long textures

What if you have a diamond/circle skin and you want the tail to end right before the next note?
![image](https://github.com/ppy/osu/assets/15729831/b1ede235-03aa-4e36-9ee6-7ae6d0467964)

Currently, this is not possible without the huge texture hack, since the tail piece is configured to always go after the body and not on top of it. If you remove support for it, you break all those skins which use the hack in a completely valid way. Can't have your cake and eat it too :)

Do I like the huge texture hack? No, that's why I proposed a different solution at #26470. But no matter what you do, if you want to support those diamond/circle skins, then percy will eventually happen, and there's no way around it.
